### DISCUSSION TITLE
Multiple instances

### DISCUSSION THREAD

**Discussion by rrangelo** [apache/netbeans#6729] [Q&A] [answered]
Hi all!

Anybody knows how can I open multiple instances of Apache NetBeans 19 and above?

Thanks in advanced!

**Comment by mbien** [comment, marked as answer]
NetBeans is built around project groups which let you switch between groups of projects within the same instance (you should be also able to open a lot of projects at the same time before things slow down). If you really need multiple instances at the same time, you have to give each instance its own config folder (analog to how firefox works for example).

```
$ netbeans --help
Usage: /home/mbien/dev/nb/platform/lib/nbexec {options} arguments

General options:
  --help                show this help 
  --jdkhome <path>      path to Java(TM) 2 SDK, Standard Edition
  -J<jvm_option>        pass <jvm_option> to JVM

  --cp:p <classpath>    prepend <classpath> to classpath
  --cp:a <classpath>    append <classpath> to classpath
WARNING: package com.apple.eio not in java.desktop
Module reload options:
  --reload /path/to/module.jar  install or reinstall a module JAR file

Additional module options:
  --modules                               
  --refresh                               Refresh all catalogs
  --list                                  Prints the list of all modules, their versions and enablement status
  --install <arg1>...<argN>               Installs provided JAR files as modules
  --disable <arg1>...<argN>               Disable modules for specified codebase names
  --enable <arg1>...<argN>                Enable modules for specified codebase names
  --update <arg1>...<argN>                Updates all or specified modules
  --update-all                            Updates all modules
  --extra-uc <arg>                        Add a extra Update Center (URL)
  --direct-disable <arg1>...<argN>        Disable module immediately
  --start-java-language-server [<arg>]    Starts the Java Language Server
  --start-java-debug-adapter-server <arg> Starts the Java Debug Adapter Server
  --open file1[:line1]...                 open specified file(s), possibly at given location; can also pass project directories
  --open-group NAME                       open a project group by shortened or full name (or unique substring)
  --close-group                           close any open project group
  --list-groups                           list available project groups

Core options:
  --laf <LaF classname> use given LookAndFeel class instead of the default
  --fontsize <size>     set the base font size of the user interface, in points
  --locale <language[:country[:variant]]> use specified locale
  --userdir <path>      use specified directory to store user settings
  --cachedir <path>     use specified directory to store user cache, must be different from userdir
  --nosplash            do not show the splash screen
```


shell 1
```
netbeans --userdir /tmp/nb_instance1
```
shell 2
```
netbeans --userdir /tmp/nb_instance2
```

Works best when the extra arguments are put into a NetBeans shortcut. If you don't set `--cachedir`, the cache will be put into the config folder.

The example above is just for demo purposes, you probably want to pick a folder which is not inside `/tmp`.

**Reply by Chris2011** [reply]
I see 2 things here.

1 with a real instance and 2 inside a new window. Just my 2 cents.

Firefox can also handle a new window as all other browsers with ctrl + n. It will not create a new instance (probably) but a new window, which is faster than bootstrapping the whole Firefox again. This is also what I would lie to see "Open project in new window" not directly in new instance. That means that NetBeans is already available and open and it just opens a new JFrame or whatever with all the other stuff inside. This will be than really fast because you don't need to open NetBeans and register modules again.

**Reply by mbien** [reply]
right, this isn't supported. NB is build around the main window + views. Context switch works over project groups. To get a new instance (main window), you would have to clone the user config. 

You can have as many editor windows open (detached) as you want, but there is only one with the main menu and views like Navigator or Projects exist only once per instance.

What I meant with the firefox comparison is that to get a new instance you have to give it its own profile which works exactly analog to the user config NB relation. its `firefox --profile` vs `netbeans --userdir`.

to open `mavenproject1` in a new NB instance you would have to do something like:
```
netbeans --userdir /tmp/nbconfig42 --open /tmp/mavenproject1
```

**Reply by neilcsmith-net** [reply]
In addition, @Chris2011, Firefox is an odd analogy as it uses lots of processes - you probably have at least one per tab.  Rewriting NetBeans to support multiple main windows and contexts in one process could be a lot of work.  Improving UI and performance of multiple processes might be worth considering.

**Reply by rma-sjn** [reply]
I solved this problem recently by writing a C/C++ program the executes my netbeans application. First it looks in the .conf file for the userdir parameter, then it looks to see if there is a lock file in the userdir folder.  The lock file signifies an instance of NB platform running.  If there is a lock file then it executes netbeans with the --userdir directory and passing a new "temporary" folder to use, othwerise I execute without the --userdir.  Don't know if that is the best solution, but it doesn't involve users editing conf files or passing in parameters and seems pretty automated.
